coap/api/transport.go

```diff
@@ -0,0 +1,191 @@
+	r.Handle("/channels/{id}/messages", gocoap.FuncHandler(receive(svc))).Methods(gocoap.POST)
+	r.Handle("/channels/{id}/messages", gocoap.FuncHandler(observe(svc))).Methods(gocoap.GET)
+	r.NotFoundHandler = gocoap.FuncHandler(NotFoundHandler)
+	return r
+}
+
+func receive(svc coap.Service) func(conn *net.UDPConn, addr *net.UDPAddr, msg *gocoap.Message) *gocoap.Message {
+	return func(conn *net.UDPConn, addr *net.UDPAddr, msg *gocoap.Message) *gocoap.Message {
+		var res *gocoap.Message
+		if msg.IsConfirmable() {
+			res = &gocoap.Message{
+				Type:      gocoap.Acknowledgement,
+				Code:      gocoap.Content,
+				MessageID: msg.MessageID,
+				Token:     msg.Token,
+				Payload:   []byte{},
+			}
+			res.SetOption(gocoap.ContentFormat, gocoap.AppJSON)
+		}
+
+		if len(msg.Payload) == 0 && msg.IsConfirmable() {
+			res.Code = gocoap.BadRequest
+			return res
+		}
+
+		cid := mux.Var(msg, "id")
+		publisher, err := authorize(msg, res, cid)
+		if err != nil {
+			res.Code = gocoap.Unauthorized
+			return res
+		}
+
+		rawMsg := mainflux.RawMessage{
+			Channel:   cid,
+			Publisher: publisher,
+			Protocol:  protocol,
+			Payload:   msg.Payload,
+		}
+
+		if err := svc.Publish(rawMsg); err != nil {
+			res.Code = gocoap.InternalServerError
+		}
+		return res
+	}
+}
+
+func observe(svc coap.Service) func(conn *net.UDPConn, addr *net.UDPAddr, msg *gocoap.Message) *gocoap.Message {
+	return func(conn *net.UDPConn, addr *net.UDPAddr, msg *gocoap.Message) *gocoap.Message {
+		var res *gocoap.Message
+		if msg.IsConfirmable() {
+			res = &gocoap.Message{
+				Type:      gocoap.Acknowledgement,
+				Code:      gocoap.Content,
+				MessageID: msg.MessageID,
+				Token:     msg.Token,
+				Payload:   []byte{},
+			}
+			res.SetOption(gocoap.ContentFormat, gocoap.AppJSON)
+		}
+
+		cid := mux.Var(msg, "id")
+		publisher, err := authorize(msg, res, cid)
+
+		if err != nil {
+			res.Code = gocoap.Unauthorized
+			return res
+		}
+
+		if value, ok := msg.Option(gocoap.Observe).(uint32); ok && value == 1 {
+			id := fmt.Sprintf("%s-%x", publisher, msg.Token)
+			if err := svc.Unsubscribe(id); err != nil {
+				res.Code = gocoap.InternalServerError
+			}
+		}
+
+		if value, ok := msg.Option(gocoap.Observe).(uint32); ok && value == 0 {
+			ch := make(chan mainflux.RawMessage)
+			id := fmt.Sprintf("%s-%x", publisher, msg.Token)
+			if err := svc.Subscribe(cid, id, ch); err != nil {
+				res.Code = gocoap.InternalServerError
+				return res
+			}
+			go handleSub(svc, id, conn, addr, msg, ch)
+			res.AddOption(gocoap.Observe, 0)
+		}
+		return res
+	}
+}
+
+func sendMessage(conn *net.UDPConn, addr *net.UDPAddr, msg *gocoap.Message) error {
+	var err error
+	now := time.Now().UnixNano() / timestamp
+	buff := new(bytes.Buffer)
+	if err = binary.Write(buff, binary.LittleEndian, now); err != nil {
+		return err
+	}
+	msg.SetOption(gocoap.Observe, buff.Bytes()[:3])
+	timeout := time.Duration(5)
+	// Try to transmit 3 times; each time duplicate timeout between attempts.
```

Comment:
Is this specified in RFC? If so, then proceed. Otherwise, consider utilizing [exponential backoff](https://github.com/cenkalti/backoff).